### PR TITLE
Add AsyncFd::with_interest

### LINKED ISSUE #3072: AsyncFd doesn't allow to decide read/write interest making  kqueue's fd fail in tokio 0.3
**Version** 0.3.2

**Platform** macOS 10.15

**Description**
Creating AsyncFd for such file description fails with:
`Os { code: 22, kind: InvalidInput, message: "Invalid argument" }`

The problem is probably in underlying mio, but I'm not sure how much changed since 0.6

The thing is PollEvented + Mio worked just fine, hence it can be considered a regression?

Just to elaborate: it is the way to create kqueue based timer, similar to linux's timer_fd

I was using `nix` to get simplified API over kqueue, but it works with 0.2's PollEvented

I use my own wrapper for timer and implement AsRawFd for it.

```rust
use libc::c_int;

pub struct RawTimer(c_int);

pub trait TimerFd: std::os::unix::io::AsRawFd + Sync + Send + Unpin {
    fn new() -> Self;
    fn set(&mut self, time: time::Duration);
    fn unset(&mut self);
    fn read(&mut self) -> usize;
}

impl std::os::unix::io::AsRawFd for RawTimer {
    #[inline(always)]
    fn as_raw_fd(&self) -> c_int {
        self.0
    }
}

impl TimerFd for RawTimer {
    fn new() -> Self {
        let fd = nix::sys::event::kqueue().unwrap_or(-1);

        //If you hit this, then most likely you run into OS imposed limit on file descriptor number
        os_assert!(fd != -1);
        Self(fd)
    }

    fn set(&mut self, time: time::Duration) {
        use nix::sys::event::*;

        let flags = EventFlag::EV_ADD | EventFlag::EV_ENABLE | EventFlag::EV_ONESHOT;
        let mut time = time.as_nanos();
        let mut unit = FilterFlag::NOTE_NSECONDS;

        if time > isize::max_value() as u128 {
            unit = FilterFlag::NOTE_USECONDS;
            time /= 1_000;
        }
        if time > isize::max_value() as u128 {
            unit = FilterFlag::empty(); // default is milliseconds
            time /= 1_000;
        }
        if time > isize::max_value() as u128 {
            unit = FilterFlag::NOTE_SECONDS;
            time /= 1_000;
        }

        let time = time as isize;
        kevent(self.0, &[KEvent::new(1, EventFilter::EVFILT_TIMER, flags, unit, time, 0)], &mut [], 0).expect("To arm timer");
    }

    fn unset(&mut self) {
        use nix::sys::event::*;

        let flags = EventFlag::EV_DELETE;
        kevent(self.0, &[KEvent::new(1, EventFilter::EVFILT_TIMER, flags, FilterFlag::empty(), 0, 0)], &mut [], 0).expect("To disarm timer");
    }

    fn read(&mut self) -> usize {
        use nix::sys::event::*;

        let mut ev = [KEvent::new(0, EventFilter::EVFILT_TIMER, EventFlag::empty(), FilterFlag::empty(), 0, 0)];

        kevent(self.0, &[], &mut ev[..], 0).expect("To execute kevent")
    }
}
```

### PR DESCRIPTION
**Pull request by HK416-is-all-you-need**
Expose `Interest` customization for `AsyncFd`

Fixes #3072

### CODE REVIEW

File: tokio/src/io/async_fd.rs
Comment (carllerche) [inline tokio/src/io/async_fd.rs:93]: Thanks, I'm wondering if this should just be `with_interest` (modeling after `Vec::with_capacity`).

The other `new_with` fns are internal and haven't received as much attention naming wise.

Thoughts @tokio-rs/maintainers ?

**Review by carllerche — COMMENTED** [review]
LGTM, I opened a minor naming point.

File: tokio/src/io/async_fd.rs
Comment (hawkw) [reply tokio/src/io/async_fd.rs:93]: i have a slight preference for `with_interest`, but i could be convinced otherwise.

Comment (HK416-is-all-you-need) [reply tokio/src/io/async_fd.rs]: Changed it to `with_interest`, it is shorter too so I would prefer it

File: tokio/src/io/async_fd.rs
Comment (Darksonn) [inline tokio/src/io/async_fd.rs:102]: We may also want to rename this to `with_handle_and_interest` for consistency, but definitely not a blocker. I will merge this shortly if nobody has any other comments.

**Review by Darksonn — APPROVED** [review]

**Pull request merged**
